### ISSUE TITLE
Embedded documents with targetDocument and discriminatorMap property do not save discriminatorField

### ISSUE THREAD

**Issue by Steveb-p** [doctrine/mongodb-odm#2203]
|    Q        |   A
|------------ | -----
| Version     | 2.1.0

### Support Question

tl;dr
When using `targetDocument` and `discriminatorMap` annotation property for `EmbedMany` (I believe `EmbedOne` as well) embedded documents do not save discriminator property.

I believe it should be mentioned in the docs that those two should not be mixed.  
If possible I would either allow `targetDocument`to coexist with `discriminatorMap` as a "sanity check" when document passed is expected to be a specific class / interface.  
Otherwise, I would throw an exception if they are used together.

Context:
I was extending an existing document to include a new type of embedded document with new properties. I had a document looking like this:
```php
use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;

/**
 * @ODM\Document
 */
class Order 
{
    /**
     * @ODM\EmbedMany(targetDocument=LegacyClass::class)
     */
    private $embedded;
}
```

then I wanted to introduce another possible target document:

```php
use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM;

/**
 * @ODM\Document
 */
class Order 
{
    /**
     * @ODM\EmbedMany(
     *     targetDocument=AbstractClass::class),
     *     strategy="atomicSetArray",
     *     discriminatorField="type",
     *     discriminatorMap={
     *         "legacy"=LegacyClass::class,
     *         "statistic_aggregate"=NewClass::class,
     *     },
     *     defaultDiscriminatorValue="legacy",
     */
    private $embedded;
}
```

However, this resulted in `type` property to never be set.

While documents inserted into this embedded collection were `NewClass` instances and have it's property, when trying to read ODM was not finding `type` property and tried to instantiate `LegacyClass`.

Therefore I think that docs should explicitly warn about mixing those two annotations (`targetDocument` & `discriminatorMap`).

**Comment by malarzm** [comment]
> If possible I would either allow targetDocumentto coexist with discriminatorMap as a "sanity check" when document passed is expected to be a specific class / interface.

IIRC we've made discriminator map strict if it's present (i.e. ODM will throw an exception if class is not listed in the map), so a sanity check is already there. My vote goes for an exception given it's not working anyway, a note in the docs will be also a nice to have.

**Comment by Steveb-p** [comment]
@malarzm so basically an additional check should be added around here:

https://github.com/doctrine/mongodb-odm/blob/master/lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php#L1929

Simply looking if `targetDocument` and `discriminatorMap` are set? And then throwing a new exception class?

If yes then I think I should be able to handle it :laughing:

**Comment by malarzm** [comment]
Exactly :)

**Comment by alcaeus** [comment]
I'll take another look at this - this should definitely work as you intend it to work.

**Comment by Steveb-p** [comment]
@alcaeus initial PR for it that I've made seems to be a little to aggresive (`isset` checks are to broad and cause exception to be thrown even then it shouldn't). Feel free to take it over and use those commits (or discard them and only use code).

**Comment by Steveb-p** [comment]
@alcaeus what's the prefered approach for this? From your comment I assume - maybe wrong - that those two annotation properties should be able to simultanously exist?
